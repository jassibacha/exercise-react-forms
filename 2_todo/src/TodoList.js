import React, { useState } from 'react';
import NewTodoForm from './NewTodoForm';
import Todo from './Todo';
import { v4 as uuid } from 'uuid';

const TodoList = () => {
    const [todos, setTodos] = useState([]);

    const renderTodos = () => {
        return (
            <div>
                {todos.map((todo) => (
                    <Todo
                        task={todo.task}
                        id={todo.id}
                        key={todo.id}
                        deleteTodo={() => deleteTodo(todo.id)}
                    />
                ))}
            </div>
        );
    };

    const deleteTodo = (todoId) => {
        // Run setBoxes again
        setTodos((todos) =>
            // Filter the boxes array from state
            todos.filter((todo) => {
                // If the ID matches, remove it!
                return todo.id !== todoId;
            })
        );
    };

    /** Add new item object to cart. */
    const addTodo = (todo) => {
        let newTodo = { ...todo, id: uuid() };
        setTodos((todos) => [...todos, newTodo]);
    };

    return (
        <div className="Todos">
            <NewTodoForm addTodo={addTodo} />
            {renderTodos()}
        </div>
    );
};

export default TodoList;
