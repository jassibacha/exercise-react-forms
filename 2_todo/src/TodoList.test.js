import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';

it('renders without crashing', function () {
    render(<TodoList />);
});

it('matches snapshot', function () {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});

it('can add a new box', function () {
    const { getByLabelText, queryByText, getByText, toHaveStyle } = render(
        <TodoList />
    );

    // check there's no Delete button
    expect(queryByText('X')).not.toBeInTheDocument();

    const taskInput = getByLabelText('Task:');
    const submitBtn = getByText('Add');

    // fill out the form
    fireEvent.change(taskInput, { target: { value: 'A Task' } });
    fireEvent.click(submitBtn);

    // box exists
    const deleteBtn = queryByText('X');
    expect(deleteBtn).toBeInTheDocument();
});

it('can delete a box', function () {
    const { getByLabelText, getByText, queryByText } = render(<TodoList />);

    const taskInput = getByLabelText('Task:');
    const submitBtn = getByText('Add');

    // fill out the form
    fireEvent.change(taskInput, { target: { value: 'A Task' } });
    fireEvent.click(submitBtn);

    // box exists
    const deleteBtn = getByText('X');
    expect(deleteBtn).toBeInTheDocument();

    // delete the box
    fireEvent.click(deleteBtn);

    // box is gone!
    expect(queryByText('X')).not.toBeInTheDocument();
});
