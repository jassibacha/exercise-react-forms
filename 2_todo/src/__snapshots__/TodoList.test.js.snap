// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Todos"
  >
    <form
      class="TodoForm"
    >
      <label
        for="task"
      >
        Task:
      </label>
      <input
        id="task"
        name="task"
        type="text"
        value=""
      />
      <button>
        Add
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
